### PR TITLE
Flap detection

### PR DESCRIPTION
**Pull request by lmars**
This is an initial attempt at a restart back-off policy when jobs crash:
- If the job is new (i.e. was created by a new / updated formation) restart it immediately and increment its restart count
- If the job was started more than 10 minutes ago, reset its restart count to zero and restart it immediately
- If the job was started less than 10 minutes ago, wait 10 \* 2 ^ (restart count - 1) minutes before restarting it

I have not added a test for this, but I have tried it out manually by switching the `backoffPeriod` to 30 seconds, manually killing jobs and seeing they are rescheduled as expected, the results of which can be seen in the following gist:

https://gist.github.com/lmars/a54a791e1434a0a311a8

Things that need addressing:
- Job restarts are not visible in the controller, so whilst the scheduler knows that jobs are queued to be restarted in the future, the controller will report that not enough jobs are running for a given formation
- When scaling down a formation, queued jobs should be removed first by just stopping their timer
- There is no upper limit on restart attempts

This depends on https://github.com/flynn/flynn-host/pull/17

### CODE REVIEW

**Comment by progrium** [comment]
What does a backoff period accomplish in this case? What state would change? It seems like it should just restart and if it fails x times, give up and mark it as broken. Something like most supervisors: http://upstart.ubuntu.com/cookbook/#respawn-limit

This is also what Heroku does. Backoff only makes sense when connecting to a remote process that itself may be restarting or recovering from load etc.

**Comment by lmars** [comment]
@progrium I have based this initial implementation on [Heroku's Dyno crash restart policy](https://devcenter.heroku.com/articles/dynos#dyno-crash-restart-policy).

It is unknown what external dependencies an app depends on, so a cool-off period may help in eventually restarting the app, but it may not (hence the need for some cut-off which I have not yet added).

**Comment by progrium** [comment]
Okay, thanks for clarifying. Trying to avoid unnecessary complexity, but this seems reasonable. Perhaps this restarting and backoff can be pushed to the host?

**Comment by titanous** [comment]
@progrium I think it is the scheduler's responsibility to implement this functionality.

**Comment by progrium** [comment]
I think the host should have a basic responsibility of trying to keep a
process running, but anything more complex than basic supervisor restart
logic should probably be in the scheduler, yes.

On Mon, Jun 2, 2014 at 11:43 AM, Jonathan Rudenberg <
notifications@github.com> wrote:

> @progrium https://github.com/progrium I think it is the scheduler's
> responsibility to implement this functionality.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/flynn/flynn-controller/pull/22#issuecomment-44861212.

## 

Jeff Lindsay
http://progrium.com

File: Godeps [hunk 1]
```diff
@@ -20,7 +20,7 @@
 		},
 		{
 			"ImportPath": "github.com/flynn/flynn-host/types",
-			"Rev": "aa072a4db1972783518e21760146411b1b4dcde9"
+			"Rev": "f3e7a4a7404d1eb8b6f042f751ce8bbe4a490ab6"
```
Comment (titanous) [inline Godeps]: This needs updating.

Comment (lmars) [inline Godeps]: Updated in e9df18a

**Comment by lmars** [comment]
@progrium 

> I think the host should have a basic responsibility of trying to keep a process running

In that case there would need to be some distinction in the host between short and long lived jobs (e.g. like [task](http://upstart.ubuntu.com/cookbook/#task-job) vs [service](http://upstart.ubuntu.com/cookbook/#service-job) Upstart jobs), but I quite like the idea of the host's job being as simple as "here is a job, run it and tell me when it's state changes"

**Comment by progrium** [comment]
Yep. It's simpler. But if you need (and have) host-level supervision
otherwise, you might as well take advantage of it? Not that it does now...
but if it did, what would you think?

On Mon, Jun 2, 2014 at 12:57 PM, Lewis Marshall notifications@github.com
wrote:

> @progrium https://github.com/progrium
> 
> I think the host should have a basic responsibility of trying to keep a
> process running
> 
> In that case there would need to be some distinction in the host between
> short and long lived jobs (e.g. like task
> http://upstart.ubuntu.com/cookbook/#task-job vs service
> http://upstart.ubuntu.com/cookbook/#service-job Upstart jobs), but I
> quite like the idea of the host's job being as simple as "here is a job,
> run it and tell me when it's state changes"
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/flynn/flynn-controller/pull/22#issuecomment-44870213.

## 

Jeff Lindsay
http://progrium.com

**Comment by lmars** [comment]
Given there are a variety of approaches to restarting crashed jobs (two already mentioned here), I think the host should stay neutral and allow clients to make there own decision on how / when to restart jobs (e.g. maybe they want to change some env vars on successive runs).

Sure if it were already implemented in the host I may of been obliged to use it, but then I may of been forced into a "try x times then mark as failed" policy rather than the exponential back off policy I have chosen to use here.

File: scheduler/main.go [hunk 2]
```diff
@@ -446,12 +453,34 @@ func (f *Formation) Rectify() {
...
+	// don't restart one-off jobs
+	if typ == "" {
+		return
+	}
 	f.mtx.Lock()
 	defer f.mtx.Unlock()
 
-	f.jobs.Remove(typ, hostID, jobID)
-	f.rectify()
+	if job := f.jobs.Get(typ, hostID, jobID); job != nil {
```
Comment (titanous) [inline scheduler/main.go]: Invert this:

``` go
job := f.jobs.Get(typ, hostID, jobID)
if job == nil {
    return
}
// ...
```

File: scheduler/main.go [hunk 3]
```diff
@@ -446,12 +453,34 @@ func (f *Formation) Rectify() {
...
+			job.restarts = 0
+		}
+
+		if job.restarts == 0 {
+			f.restart(typ, hostID, jobID, 0)
+		} else {
+			// wait backoffPeriod * 2 ^ (restarts - 1) before restarting
+			duration := backoffPeriod
+			for i := 0; i < job.restarts-1; i++ {
+				duration = duration * 2
```
Comment (titanous) [inline scheduler/main.go]: I think `duration *= 2` should work.

File: scheduler/main.go [hunk 4]
```diff
@@ -489,45 +519,69 @@ func (f *Formation) add(n int, name string) {
...
 
-		g.Log(grohl.Data{"host.id": h.ID, "job.id": config.ID})
+func (f *Formation) restart(typ, hostID, jobID string, restarts int) error {
+	g := grohl.NewContext(grohl.Data{"fn": "restart", "app.id": f.AppID, "release.id": f.Release.ID})
+	g.Log(grohl.Data{"old.host.id": hostID, "old.job.id": jobID})
 
-		job := f.jobs.Add(name, h.ID, config.ID)
-		job.Formation = f
-		f.c.jobs.Add(h.ID, config.ID, job)
+	hID, jID, err := f.start(typ, nil)
```
Comment (titanous) [inline scheduler/main.go]: I'd prefer to see `new` and `old` prefixes to the variable names, I think it makes this a bit clearer.

File: scheduler/main.go [hunk 5]
```diff
@@ -489,45 +519,69 @@ func (f *Formation) add(n int, name string) {
...
+	if err != nil {
+		return err
+	}
+	g.Log(grohl.Data{"new.host.id": hID, "new.job.id": jID})
+	f.jobs.Remove(typ, hostID, jobID)
+	j := f.jobs.Add(typ, hID, jID)
+	j.Formation = f
+	j.restarts = restarts + 1
```
Comment (titanous) [inline scheduler/main.go]:
``` go
j.restarts++
```

Comment (titanous) [inline scheduler/main.go]: Or not I guess?

Comment (titanous) [inline scheduler/main.go]: Looking at `RestartJob`, it seems like the `restarts` argument to this method is redundant, am I missing something?

**Comment by titanous** [comment]
@lmars Anything left before this can be merged?

**Comment by titanous** [comment]
How hard would it be to add one or more tests for this?

File: scheduler/main.go (hunk 5 above)
Comment (lmars) [inline scheduler/main.go]: So I was passing in `restarts` as here there was no indication of how many restarts the stopped job has previously.

I have now updated this function to get the stopped job as an argument, which I think is clearer: 021dc6f

**Comment by lmars** [comment]
@titanous 

> How hard would it be to add one or more tests for this?

Probably not too hard, I could add a channel to the context which receives events when jobs are restarted, and then check the correct amount of time has elapsed, but with being time based I'll just need to make sure its predictable. I'll give it a go.

**Comment by titanous** [comment]
> with being time based I'll just need to make sure its predictable

One option is to mock out `time.AfterFunc` using a private global.

**Comment by lmars** [comment]
@titanous 

> One option is to mock out `time.AfterFunc` using a private global.

How would I go about doing that?

**Comment by titanous** [comment]
@lmars 

Somewhere in `scheduler.go`:

``` go
var timeAfterFunc = time.AfterFunc
```

And use `timeAfterFunc` everywhere instead of `time.AfterFunc`. Then in your tests you can assign `timeAfterFunc` to a function of your choice.

**Comment by lmars** [comment]
@titanous I have added a test (04d4f4b), along with some refactoring of the event simulation (3d18e5b)

File: testutils/fake_cluster.go [hunk 6]
```diff
@@ -56,15 +57,27 @@ func (c *FakeCluster) AddJobs(req *host.AddJobsReq) (*host.AddJobsRes, error) {
...
 		}
+		if client, ok := c.hostClients[hostID]; ok {
+			for _, job := range jobs {
+				// Send the start event asynchronously after a short period of time
+				// to give the caller chance to finish processing previous events
+				// before it receives the start event (e.g. when restarting jobs,
+				// the scheduler needs to track the new hostID and jobID before it
+				// can deal with a start event for that job).
+				go func() {
+					time.Sleep(100 * time.Millisecond)
```
Comment (titanous) [inline testutils/fake_cluster.go]: Will this sleep affect test success if the runner is slow?

Comment (lmars) [inline testutils/fake_cluster.go]: I was thinking about this too. What needs to happen is the `AddJobs` function must return _before_ the start event is sent, and although this makes that a likely scenario, I guess it's not really a guarantee. Perhaps I should look at making all callers of this method immune to the issue? WDYT?

Comment (titanous) [inline testutils/fake_cluster.go]: Yeah, that seems like racy behavior on the part of the consumers of `AddJobs`. The correct way to handle this is to subscribe to the event stream before calling `AddJobs`.

Comment (lmars) [inline testutils/fake_cluster.go]: Ok I have dealt with this in 011f5e6 by ensuring that the scheduler tracks jobs _before_ making the request to the host to start the job, removing it if there is an error (this is [how it currently works](https://github.com/flynn/flynn-controller/blob/master/scheduler/main.go#L520), so my headache was self inflicted!)

File: scheduler/main.go [hunk 7]
```diff
@@ -213,6 +218,11 @@ func (c *context) watchHost(id string, events chan<- *host.Event) {
 
 	g := grohl.NewContext(grohl.Data{"fn": "watchHost", "host.id": id})
 
+	// Empty event to mark the start of watching a host
+	if events != nil {
+		events <- &host.Event{}
```
Comment (titanous) [inline scheduler/main.go]: Could use `nil` as the sentinel.

Comment (lmars) [inline scheduler/main.go]: Personally I don't like passing `nil` around, but if thats idiomatic then I will change

Comment (titanous) [inline scheduler/main.go]: Neither is more idiomatic, it just seems easier to check than testing for a zero `Event`.

Comment (lmars) [inline scheduler/main.go]: true: aefc2e5

**Comment by titanous** [comment]
Awesome!

**Comment by miracle2k** [comment]
I've always wondered how you were intending to keep jobs up, but I think I'm with @progrium here, and my reason is this: flynn-controller is a very high-level and specific API. I do not want to use this specific Heroku-like system (but instead say, use a template-based approach ala meastro-ng), but now, rather than just generating simple job descriptions (which wouldn't even require a persistent daemon), I have to re-implement something as essential and crucial as keeping jobs running.

That is a whole different concern than the specific design of a "config set" API, and it seems like your binding the two things together here.

It's my understanding that I can give a job to CoreOS fleet, and it will ensure that the job keeps running, even if a host goes down. I'd like a flynn-host cluster to to be able to do the same.

**Comment by titanous** [comment]
@miracle2k In order to handle host outages without a scheduler daemon, the flynn-host leader would need to make scheduling decisions. We decided very early on that it should be a "scheduling framework" and do as little as possible. Currently only the scheduler that started the job has enough information to choose another host to run it on. Adding enough logic and metadata to do this in flynn-host either shifts scheduling decisions to flynn-host or duplicates work.

**Comment by titanous** [comment]
@miracle2k I think the solution is a more generic low-level scheduler daemon that can handle stuff like this.

**Comment by miracle2k** [comment]
@titanous I would have suggested that as an alternative.
